qa/L0_trace/opentelemetry_unittest.py

```diff
@@ -153,121 +191,560 @@ def _check_events(self, span_name, events):
+                "bls_simple": ["compute", "ensemble_add_sub_int32_int32_int32"],
+                "ensemble_add_sub_int32_int32_int32": ["simple"],
+                "simple": ["compute"],
+            }
+        )
+        for key in expected_parent_span_dict.keys():
+            expected_parent_span_dict[key].sort()
+
+        self._test_trace(
+            headers=headers,
+            expected_number_of_spans=expected_number_of_spans,
+            expected_counts=expected_counts,
+            expected_parent_span_dict=expected_parent_span_dict,
+        )
 
-    inputs = prepare_data(grpcclient)
-    triton_client_grpc.infer("simple", inputs)
+    def _test_ensemble_trace(self, headers=None):
+        """
+        Helper function, that specifies expected parameters to evaluate trace,
+        collected from running 1 inference request for an
+        `ensemble_add_sub_int32_int32_int32` model.
+        """
+        expected_number_of_spans = 4
+        expected_counts = dict(
+            {
+                "compute": 1,
+                self.simple_model_name: 1,
+                self.ensemble_model_name: 1,
+                self.root_span: 1,
+            }
+        )
+        expected_parent_span_dict = dict(
+            {
+                "InferRequest": ["ensemble_add_sub_int32_int32_int32"],
+                "ensemble_add_sub_int32_int32_int32": ["simple"],
+                "simple": ["compute"],
+            }
+        )
+        for key in expected_parent_span_dict.keys():
+            expected_parent_span_dict[key].sort()
+
+        self._test_trace(
+            headers=headers,
+            expected_number_of_spans=expected_number_of_spans,
+            expected_counts=expected_counts,
+            expected_parent_span_dict=expected_parent_span_dict,
+        )
 
-    inputs = prepare_data(httpclient)
-    triton_client_http.infer("ensemble_add_sub_int32_int32_int32", inputs)
+    def test_http_trace_simple_model(self):
+        """
+        Tests trace, collected from executing one inference request
+        for a `simple` model and HTTP client.
+        """
+        triton_client_http = httpclient.InferenceServerClient(
+            "localhost:8000", verbose=True
+        )
+        inputs = prepare_data(httpclient)
+        triton_client_http.infer(self.simple_model_name, inputs)
+
+        self._test_simple_trace()
+
+    def test_http_trace_simple_model_context_propagation(self):
+        """
+        Tests trace, collected from executing one inference request
+        for a `simple` model, HTTP client and context propagation,
+        i.e. client specifies OTel headers, defined in `self.client_headers`.
+        """
+        triton_client_http = httpclient.InferenceServerClient(
+            "localhost:8000", verbose=True
+        )
+        inputs = prepare_data(httpclient)
+        triton_client_http.infer(
+            self.simple_model_name, inputs, headers=self.client_headers
+        )
```

Comment:
**This is totally optional "follow-up PR" material** if you like the idea, but I think a lot of these functions for:
1. http/grpc
2. with/without context propogation
3. simple/ensemble/bls

Could be condensed a lot with something like `unittest.subtest` or `parameterized`:
```python
def test_trace(self):
  clients = [httpclient, grpcclient]
  header_list = [None, self._client_headers]
  models = self.test_models # [self.simple_model_name, self.ensemble, self.bls ...]

  permutations = itertools.product(clients, header_list, models)
  for p in permutations:
    client_type, headers, model = p[0], p[1], p[2]
    with self.subtest(permutation=p):
        client = client_type.InferenceServerClient(...) # can probably be smart and include url somewhere
        inputs = prepare_data(client_type) 
        client.infer(model, inputs, headers)
        self._test_trace(model, headers=headers)
```

Doesn't have to be exactly like this, but seems like there's room for cleanup or re-using logic